DSpace/DSpace
# ROR Integration - OAI PMH & Orcid

**Please note that this development has been funded by the University of California - California Digital Library.**

## References
* This PR builds on top of DSpace/Dspace#9238
* Related to Dspace/dspace-angular#2719

## Description
This feature introduces ROI metadata visualization inside the OAI-PMH and its synchronization with ORCID profile, if configured any.

## Instructions for Reviewers
First of all, you need to enable the [OAI-PMH](https://wiki.lyrasis.org/display/DSDOC7x/OAI) and [ORCID synchronization](https://wiki.lyrasis.org/display/DSDOC7x/ORCID+Integration#ORCIDIntegration-Overview) feature. 

### OAI Integration
> You can find everything needed in these guides: [OAI - DSpace 7.x Documentation - LYRASIS Wiki](https://wiki.lyrasis.org/display/DSDOC7x/OAI) and [OpenAIRE4 Guidelines Compliancy - DSpace 7.x Documentation - LYRASIS Wiki](https://wiki.lyrasis.org/display/DSDOC7x/OpenAIRE4+Guidelines+Compliancy) 
Configuration steps:
1. Enable the OAI server inside oai.cfg:
```
2. Enable (true) or disable (false) OAI-PMH server
oai.enabled = true
# When enabled, OAI-PMH server is available at this path
oai.path = oai
```
3. Configure some collections to use the standard OpenAIRE form support inside item-submission.xml:
```xml
<name-map collection-handle="123456789/2" submission-name="openAIREPublicationSubmission" />
<name-map collection-handle="123456789/3" submission-name="openAIREPersonSubmission" />
<name-map collection-handle="123456789/5" submission-name="openAIREProjectSubmission" />
<name-map collection-handle="123456789/4" submission-name="openAIREOrganizationSubmission" />
```
4. Initialize all the relationships and metadata related to the openaire integration:
```
[/dspace]/bin/dspace initialize-entities -f [/dspace]/config/entities/openaire4-relationships.xml
[/dspace]/bin/dspace registry-loader -metadata [/dspace]/config/registries/openaire4-types.xml
```
5. Enable OAI interface:
```
oai.enabled = true
```
6. Replace the file `[/dspace/]config/spring/api/virtual-metadata.xml` with the one named `[/dspace/]config/spring/api/virtual-metadata.xml.openaire` by renaming this one, in order to see all the additional openaire information inside the OAI-PMH interface.
7. Just run the command:
```
[/dspace]/bin/dspace oai import -c
```
To reinitialize the oai solr core with your configuration.

### ORCID Synchronization
> You can find everything needed in this guide: [ORCID Integration - DSpace 7.x Documentation - LYRASIS Wiki](https://wiki.lyrasis.org/display/DSDOC7x/ORCID+Integration#ORCIDIntegration-Overview) 

1. Configure orcid API integration: 
```
# These URLs are for testing against ORCID's Sandbox API
# These are only useful for testing, and you must first request a Sandbox API Key from ORCID
orcid.domain-url= https://sandbox.orcid.org
orcid.api-url = https://api.sandbox.orcid.org/v3.0
orcid.public-url = https://pub.sandbox.orcid.org/v3.0
# Keep in mind, these API keys MUST be for the Sandbox API if you use "sandbox.orcid.org" URLs above!
orcid.application-client-id = <YOUR-SANDBOX-ORCID-CLIENT-ID>
orcid.application-client-secret = <YOUR-SANDBOX-ORCID-CLIENT-SECRET>
Enable orcid login feature - plugin.sequence.org.dspace.authenticate.AuthenticationMethod = org.dspace.authenticate.OrcidAuthentication
```
2. Enable orcid synchronization
```
orcid.synchronization-enabled = true
```
3. Enable orcid synchronization consumer (orcidqueue)
```
event.dispatcher.default.consumers = versioning, discovery, eperson, orcidqueue
```
4. Enable orcid validation:
```
orcid.validation.work.enabled = true
orcid.validation.funding.enabled = true
```
5. Configure new ROR identifier mapping:
```
orcid.mapping.organization.identifiers = organization.identifier.ror::ROR
orcid.validation.organization.identifier-sources = ROR
```

List of changes in this PR:
* Changes coming from the live-import feature of ROR
* Changes involving configuration files to enable the new features

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

### OAI Integration
| Step  | Screenshot |
| ------- | ---------------- |
|Import an `Organization` using the ROR live import feature| ![image](https://github.com/DSpace/DSpace/assets/107129187/be165647-db61-49be-a1ae-1fea141e38ef) |
| Link a `Project` to an `Organization` entity | ![image](https://github.com/DSpace/DSpace/assets/107129187/bd8fcb50-02b3-4c01-95b6-1ebef1a477d8) |
| Create a `Publication` with the OpenAIRE form configuration and link it to the `Project`| ![image](https://github.com/DSpace/DSpace/assets/107129187/8fa16765-6e00-4afa-b497-cfc51b630846) |
| Update the OAI index | `[/dspace]/bin/dspace oai import -c`|
| Visualize the entity inside the `oai` interface (`[dspace-url]/server/oai/...`) | ![image](https://github.com/DSpace/DSpace/assets/107129187/b5e3d0d8-654b-4fbb-8456-802ade8258fc)|

### ORCID Synchronization
To verify the OAI integration, you need to:
| Step | Screenshot |
| ------ | ----------------- |
| Login with an ORCID user that has a Researcher Profile and that is able to submit |![image](https://github.com/DSpace/DSpace/assets/107129187/ea8854f5-9165-480a-bd4b-95e115484b1c) |
| Enable ORCID Synchronization from its profile page | ![image](https://github.com/DSpace/DSpace/assets/107129187/e7bc9f56-edbe-4fb4-b317-24aa4eafec7a)|
| Import a ROR `Organization` with the live import feature | ![image](https://github.com/DSpace/DSpace/assets/107129187/af496c3b-f602-4dd5-a1af-27cfd6d4ca13)|
| Link a `Project` to the `Organization` and the ORCID User | ![image](https://github.com/DSpace/DSpace/assets/107129187/617de915-2bcd-403b-8065-b3de8af36255)|
| You should have a new `Project` entry inside the ORCID Synchronization Queue of your profile | ![image](https://github.com/DSpace/DSpace/assets/107129187/2f8c4946-de7d-46b1-b02b-140590502eb1) |
| Synchronize the change to ORCID, and you'll see the new Funding with the related ROR identifier | ![image](https://github.com/DSpace/DSpace/assets/107129187/6821652c-7e77-4b51-a955-46848bad4986) ![image](https://github.com/DSpace/DSpace/assets/107129187/6b7f0b55-f3e2-4d84-be14-ab8b95d1ee51)|

## Checklist
